File: NumberSelection/Scala/src/main/scala/GuessNumber.scala
object GuessNumber {
  def guess(selectedNumber: Int, input: () => Int, output: String => Unit, attempts:Int = 1): Boolean = {
    val responses = Array("Aim lower", s"You guessed it in $attempts ${if (attempts == 1) "attempt" else "attempts"}", "Aim higher")
    
    val guessMatch = (selectedNumber compare input()).sign
    
    output(responses(guessMatch + 1))
    
    guessMatch == 0
  }
  
  def play(selectedNumber: Int, input:() => Int, output: String => Unit): Unit ={
    Iterator.from(1).find( attempts => guess(selectedNumber, input, output, attempts) )
  }
}

File: NumberSelection/Ruby/lib/guess_number.rb
def guess(selected_number, input, output, attempts = 1)
  responses = ['Aim lower', "You guessed it in #{attempts} " + (attempts == 1 ? 'attempt' : 'attempts'), 'Aim higher']
  guess_match = selected_number <=> input.call

  output.call(responses[guess_match + 1])
  
  guess_match.zero?
end

def play(selected_number, input, output)
  (1..).find { |attempts| guess(selected_number, input, output, attempts) } 
end
